# Retrieval Augmented Generation (RAG)

We can think of large language models (LLMs) as people who have read every book in a library up until a certain date. 📚 They've remembered the content with varying degrees of accuracy, but their knowledge base consists of everything that was in that library.

In the case of LLMs, the "library" is actually a massive collection of internet documents—heavily filtered for quality—along with some curated private datasets, such as newspaper archives and academic papers.

If we need the model to understand or use information that was not (or only barely) included in its training data, we have to provide that information in the prompt. However, there's a limit to how much information we can give at once. This is due to something called a context window—a maximum number of tokens (e.g., 8,096 tokens) that the model can process in a single prompt.

## Augmented

In [2]:
from google import genai
from IPython.display import Markdown, display
from dotenv import load_dotenv, find_dotenv

load_dotenv(find_dotenv(), override=True)
client = genai.Client()
response = client.models.generate_content(
    model="gemini-3-flash-preview", contents="Hey, I am looking for some gifts, could you recommend some"
)

display(Markdown(response.text))

I would love to help you find the perfect gift! To give you the best recommendations, I’ve broken some ideas down by "personality type" below.

If you can tell me a bit more about **who** you are buying for, your **budget**, and the **occasion**, I can give you a much more specific list.

In the meantime, here are some "crowd-pleaser" ideas:

### 1. For the "Homebody" (Comfort & Cozy)
*   **Weighted Blanket:** Great for relaxation and better sleep.
*   **High-End Candle:** Brands like *Diptyque, Boy Smells,* or *Voluspa* feel like a luxury treat.
*   **Luxury Loungewear:** A high-quality silk robe or a set of soft bamboo pajamas.
*   **Electric Mug Warmer:** Perfect for people who take a long time to drink their coffee or tea.

### 2. For the Tech & Gadget Lover
*   **Portable Power Bank:** A high-capacity one (like Anker) is always useful.
*   **AirTag or Tile Starters:** For the person who is always losing their keys or wallet.
*   **Sunrise Alarm Clock:** Gently wakes you up with light instead of a loud buzzing sound.
*   **Noise-Canceling Headphones:** Great for commuters or people who work from home.

### 3. For the Foodie or Drink Enthusiast
*   **Hot Sauce Sampler:** A fun way to try different spice levels.
*   **Cold Brew Coffee Maker:** Simple, practical, and saves money on Starbucks runs.
*   **Artisanal Olive Oil/Balsamic Set:** Something they wouldn't usually buy for themselves.
*   **Cocktail/Mocktail Kit:** Includes syrups, bitters, and a nice shaker.

### 4. For the "Experience" Seeker
*   **MasterClass Subscription:** Online classes taught by celebrities/experts (cooking, writing, etc.).
*   **Museum or Botanical Garden Membership:** A gift that lasts a whole year.
*   **Gift Card to a Local Bookstore:** Pair it with a nice bookmark.
*   **Cooking Class Voucher:** A fun activity they can do with a friend or partner.

### 5. Unique & Personalized
*   **Custom Illustration:** A digital drawing of their pet or their house (many artists on Etsy do this).
*   **Personalized Stationery:** A set of nice note cards with their name on them.
*   **The "New York Times" Birthday Book:** A collection of the front pages from every year since they were born.

---

**To help me narrow this down, could you tell me:**
1.  **Who is it for?** (Partner, parent, friend, boss?)
2.  **What is the occasion?** (Birthday, anniversary, housewarming, "thank you"?)
3.  **What is your budget?** (e.g., Under $25, around $100, or "sky is the limit"?)
4.  **What are one or two things they love?** (e.g., Gardening, Star Wars, hiking, organization?)

---

Let's now do the same with the context that it is currently Easter and see how the model adapts its gift recommendations.

In [7]:
extra_content = "It is Easter 2025."

chat_response = client.models.generate_content(model="gemini-3-flash-preview", contents='Hey, I am looking for some gifts, could you recommend some' + extra_content
)

display(Markdown(chat_response.text))

Easter 2025 falls on **Sunday, April 20th**. Since it’s a later Easter, we can expect warmer spring weather, which opens up some great options for outdoor-themed gifts alongside the traditional treats.

Here are some gift recommendations broken down by category:

### 1. For Children (Beyond just candy)
*   **The "Outdoor Fun" Kit:** Since it will be late April, think about items for the yard. Bubble machines, colorful kites, sidewalk chalk, or a high-quality jump rope.
*   **LEGO Easter Sets:** LEGO usually releases seasonal sets like the "Easter Rabbit" or "Spring Animal Playground." They are great for keeping kids quiet and focused after the egg hunt.
*   **Personalized Storybooks:** Books where the child is the main character helping the Easter Bunny.
*   **Gardening Set:** A "My First Garden" kit with seeds (sunflowers or zinnias), a small trowel, and a watering can.
*   **Plush Toys:** You can’t go wrong with a classic **Jellycat bunny**—they are incredibly soft and come in various spring colors.

### 2. For Teens & Young Adults
*   **Self-Care "Basket":** Instead of a traditional basket, use a trendy makeup bag and fill it with floral-scented face masks, lip balms (like Laneige or Summer Fridays), and silk hair scrunchies.
*   **Tech Accessories:** Pastel-colored AirPod cases, a portable power bank, or a new phone charm.
*   **Instax Mini Camera & Film:** Perfect for capturing family moments during the Easter brunch.
*   **Crocs & Jibbitz:** Pastel-colored Crocs with spring-themed charms (flowers, bunnies, suns) are still very much in style.

### 3. For Adults / Hostess Gifts
*   **Gourmet Food Basket:** Move away from cheap chocolate. Think artisanal honey, high-end olive oil, locally roasted coffee, or a box of luxury macarons (like Ladurée).
*   **Spring Home Scents:** A high-end candle from brands like **NEST (Grapefruit or Bamboo)** or **Voluspa**.
*   **Live Plants:** A potted hydrangea, a classic lily, or a beautiful orchid in a ceramic pot is more lasting than a bouquet.
*   **Linens:** Set of embroidered linen cocktail napkins with spring motifs (birds, eggs, or wildflowers).

### 4. Meaningful & Religious Gifts
*   **Willow Tree Figurines:** These hand-sculpted figures often represent themes of hope and renewal.
*   **Jewelry:** A delicate silver cross necklace or a "Mustard Seed" charm.
*   **Devotional Journals:** A high-quality leather-bound journal for reflection.

### 5. Unique "Experience" Gifts
*   **Butterfly Growing Kit:** A hit for families; you get a voucher for live caterpillars and watch them turn into Painted Lady butterflies to release in late April.
*   **Tickets to a Botanical Garden:** Perfect for a post-Easter Sunday stroll.
*   **Cooking Class:** A spring-themed baking class (learning to make macarons or hot cross buns).

---

### Tips for 2025 Trends:
*   **The "Sustainable" Basket:** Use a reusable felt or wicker basket that can be used for storage later, and look for eco-friendly "grass" made of paper rather than plastic.
*   **Color Palette:** For 2025, look for "Butter Yellow," "Peach Fuzz" (2024's color of the year still trending), and "Soft Sage Green."

**To give you a better recommendation, let me know:**
1.  Who are you shopping for? (Kids, partner, parents?)
2.  What is your budget per gift?
3.  Are you looking for something traditional or something more modern/unique?

---

The notebook is named **RAG (Retrieval-Augmented Generation)**. In this context, "augmentation" refers to enriching the generation process with additional, task-specific information. However, when the volume of available data is too large to insert in full, we must **retrieve** only the most relevant parts to use as context.

**Note**: Having a context window of for example 128.000 tokens does not mean that we should fill it up until that. Adding more and more tokens leads to more latency (i.e. it takes longer for the model to serve the user) and there could be also an information overload i.e. that the information fits into the context window but there is just too much information for the model may overlook important details. (This in the end could be only found out by testing, but personally I think in terms of information density and complexity. The german "Bild Zeitung" has for example a very low information density and complexity while a medical textbook has a very high information density and complexity. Hence doing Q&A over 30.000 tokens of Bild articles is definitively not the same as doing Q&A over 30.000 tokens of medical textbook pages.

## Retrieval

For actually retrieving something we need data. Let us consider for now this simplified database of ten different entries/documents about talks given on a future festival. We want to build a small RAG chatbot for festival visitors where they can ask for their specific interests what talks would fits the most. 

**Note** The dataset is so small that I would personally just completly put it always in the context but for demonstration purposes let's use the talks as single documents. Read through them to get a grasp of the data below:

##### 1. 🌿 **The Next Nature: Designing with the Future in Mind**  
*Explore how biomimicry, regenerative design, and synthetic biology are reshaping architecture, materials, and cities. Where does design end and nature begin?*

---

##### 2. 🧠 **Neural Frontiers: The Brain-Computer Interface Revolution**  
*From thought-controlled devices to memory enhancement, dive into the fast-evolving world of brain tech — and the ethical mazes it brings.*

---

##### 3. 🎨🤖 **AI & the Soul of Art: Who Really Owns Creativity?**  
*Artists, coders, and philosophers debate the rise of generative art. Can machines make meaning? And where does human intuition still reign supreme?*

---

##### 4. 💼🚫 **Post-Work: Imagining a Life Beyond Jobs**  
*As automation reshapes labor, what comes next? UBI, digital nomadism, reputation economies — a candid discussion about freedom, purpose, and survival.*

---

##### 5. 🪐 **Planet B: Terraforming Ideas for Earth 2.0**  
*Science fiction meets climate urgency. This talk blends real research with wild speculation — from Mars domes to floating cities in the clouds of Venus.*

---

##### 6. 🕶️🌐 **Reality is Optional: The Rise of Immersive Worlds**  
*VR, AR, XR — and whatever’s next. What happens when our digital spaces feel more real than reality itself? And who gets to write the rules?*

---

##### 7. 🧬⏳ **The Ethics of Immortality: Living Forever in a Mortal World**  
*Cryonics, gene editing, mind uploading — tech is chasing eternal life. But what would it mean for love, loss, and the human condition?*

---

##### 8. 💻⚖️ **Code as Culture: Programming the Future We Want**  
*Code is not neutral — it shapes societies. This talk explores how developers are becoming the new lawmakers, and how we hold them accountable.*

---

##### 9. 🕸️🛠️ **The Wild Web: Reclaiming the Internet from Algorithms**  
*Can we rebuild the web for people, not profit? Meet the rebels, hackers, and dreamers creating decentralized, community-first digital spaces.*

---

##### 10. 🕰️🚀 **Time Travelers Welcome: Building the Long Now**  
*Futurists, historians, and deep-time thinkers gather to explore projects that think in centuries — from 10,000-year clocks to interstellar archives.*


---

Okay, we see depending on your interests there might different talks to be the most interesting ones.

First we need to prepare the talks 

In [14]:
festival_talks = [
    "🌿 The Next Nature: Designing with the Future in Mind — Explore how biomimicry, regenerative design, and synthetic biology are reshaping architecture, materials, and cities. Where does design end and nature begin?",
    "🧠 Neural Frontiers: The Brain-Computer Interface Revolution — From thought-controlled devices to memory enhancement, dive into the fast-evolving world of brain tech — and the ethical mazes it brings.",
    "🎨🤖 AI & the Soul of Art: Who Really Owns Creativity? — Artists, coders, and philosophers debate the rise of generative art. Can machines make meaning? And where does human intuition still reign supreme?",
    "💼🚫 Post-Work: Imagining a Life Beyond Jobs — As automation reshapes labor, what comes next? UBI, digital nomadism, reputation economies — a candid discussion about freedom, purpose, and survival.",
    "🪐 Planet B: Terraforming Ideas for Earth 2.0 — Science fiction meets climate urgency. This talk blends real research with wild speculation — from Mars domes to floating cities in the clouds of Venus.",
    "🕶️🌐 Reality is Optional: The Rise of Immersive Worlds — VR, AR, XR — and whatever’s next. What happens when our digital spaces feel more real than reality itself? And who gets to write the rules?",
    "🧬⏳ The Ethics of Immortality: Living Forever in a Mortal World — Cryonics, gene editing, mind uploading — tech is chasing eternal life. But what would it mean for love, loss, and the human condition?",
    "💻⚖️ Code as Culture: Programming the Future We Want — Code is not neutral — it shapes societies. This talk explores how developers are becoming the new lawmakers, and how we hold them accountable.",
    "🕸️🛠️ The Wild Web: Reclaiming the Internet from Algorithms — Can we rebuild the web for people, not profit? Meet the rebels, hackers, and dreamers creating decentralized, community-first digital spaces.",
    "🕰️🚀 Time Travelers Welcome: Building the Long Now — Futurists, historians, and deep-time thinkers gather to explore projects that think in centuries — from 10,000-year clocks to interstellar archives."
]


For the retrieval part we need a kind of search based on the question over talks. The standard way was to do a key word search but with the upcoming AI era that default changed to a vector search. Therefor we need another type of model: An embedding model. An embedding model takes a text and transforms it into a vector of a certain size (e.g. 1*1048). Every text is mapped to a the same vector size. The embedding models are pretrained that they map similar texts close to each other in the vector space (so LLMs aren't the only cool models). Luckily ollama and also all major providers like openai and huggingface also provide embedding models. We will now take the currently best embedding models from ollama (lukily the embedding models are usually really small < 1GB)

Run **ollama pull nomic-embed-text** in your terminal

Ollama is able to serve the LLM and the embedding model at the same time (given enough (V)RAM)

In [11]:
# Generate an embedding for a single input
response = client.models.embed_content(model='gemini-embedding-2', contents='opencampus is the best!')

# Access the embedding vector
embedding = response.embeddings
print(embedding)
print(len(embedding[0].values))

[ContentEmbedding(
  values=[
    0.003136372,
    0.013906244,
    -0.009642321,
    0.02467294,
    0.008548629,
    <... 3067 more items ...>,
  ]
)]
3072


As we can see the text is converted into a vector with a size 1*3072

In [25]:
# Generate an embedding for a batch input
all_embeddings: list[list[float]] = []

for talk in festival_talks:
    resp = client.models.embed_content(model='gemini-embedding-2', contents=talk)

    all_embeddings.append(resp.embeddings[0].values)

In [27]:
# Access the embedding vectors
print(len(all_embeddings), len(all_embeddings[0]), len(all_embeddings[1]))

10 3072 3072


So we can see now that we have a list of 10 entries each with a vector with 3072 entries

In [7]:
#pip install numpy

In [31]:
import numpy as np

vectors = np.array(all_embeddings)

query = np.array(client.models.embed_content(model='gemini-embedding-2', contents='I am interested in AI talks').embeddings[0].values)

# Compute Euclidean distances
differences = vectors - query  # shape: (n_vectors, vector_dim)
distances = np.linalg.norm(differences, axis=1)

# Index of closest vector
closest_index = np.argmin(distances)

# Sorted indices by distance
sorted_indices = np.argsort(distances)

print("Closest vector index:", closest_index)
print("Distance:", distances[closest_index])
print("Sorted indices by distance:", sorted_indices)


Closest vector index: 7
Distance: 0.8137546815583194
Sorted indices by distance: [7 9 2 1 4 5 0 6 8 3]


In [32]:
festival_talks[closest_index]

'💻⚖️ Code as Culture: Programming the Future We Want — Code is not neutral — it shapes societies. This talk explores how developers are becoming the new lawmakers, and how we hold them accountable.'

In [33]:
# Sort festival_talks according to the sorted_indices
sorted_festival_talks = [festival_talks[i] for i in sorted_indices]

# Optionally print it
for num, talk in enumerate(sorted_festival_talks):
    print(num+1, talk)
    print()

1 💻⚖️ Code as Culture: Programming the Future We Want — Code is not neutral — it shapes societies. This talk explores how developers are becoming the new lawmakers, and how we hold them accountable.

2 🕰️🚀 Time Travelers Welcome: Building the Long Now — Futurists, historians, and deep-time thinkers gather to explore projects that think in centuries — from 10,000-year clocks to interstellar archives.

3 🎨🤖 AI & the Soul of Art: Who Really Owns Creativity? — Artists, coders, and philosophers debate the rise of generative art. Can machines make meaning? And where does human intuition still reign supreme?

4 🧠 Neural Frontiers: The Brain-Computer Interface Revolution — From thought-controlled devices to memory enhancement, dive into the fast-evolving world of brain tech — and the ethical mazes it brings.

5 🪐 Planet B: Terraforming Ideas for Earth 2.0 — Science fiction meets climate urgency. This talk blends real research with wild speculation — from Mars domes to floating cities in the cl

So for real retrieval we have to set a retrieval count of how many documents we want to retrieve e.g. **k=3**

In [34]:
retrieved_festival_talks = sorted_festival_talks[:3]

Let's now make our final call

In [36]:
retrieved_content = '\n\n'.join(retrieved_festival_talks)

chat_response = client.models.generate_content(model="gemini-3-flash-preview", 
                                               contents='User Question:\nI am interested in AI talks\n\n' + 
                                               'Answer the user question on festival talks as best as you can based on the retrieved festival talks\n\nRetrieved content:\n\n' + 
                                               retrieved_content)

display(Markdown(chat_response.text))

Based on the retrieved festival talks, there is one session specifically focused on AI, as well as a related talk on the societal impact of programming:

*   **AI & the Soul of Art: Who Really Owns Creativity?**
    This talk features a debate between artists, coders, and philosophers regarding the rise of generative art. It explores whether machines can create meaning and discusses where human intuition still holds value in the creative process.

*   **Code as Culture: Programming the Future We Want**
    While not exclusively about AI, this session is highly relevant as it explores how code (which powers AI) shapes our society and discusses the accountability of the developers who act as the "new lawmakers."

## Conclusion

Okay we saw that the whole idea of retrieval augmented generation is quite simple and depends on augmenting the LLM with extra context in the prompt. The context could also be retrieved if there is a lot of content and we are not able to just stick everything in the prompt. Now we already have a nice RAG chatbot coded by hand!!